### PR TITLE
Fix build error from the bottom - part 1

### PR DESCRIPTION
**Pull request by brycewang-microsoft**

### CODE REVIEW

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/IModuleClient.cs [hunk 1]
```diff
@@ -22,9 +21,9 @@ public interface IModuleClient
 
         Task SetDefaultMethodHandlerAsync(MethodCallback callback);
 
-        Task<Twin> GetTwinAsync();
+        Task<TwinProperties> GetTwinAsync();
 
-        Task UpdateReportedPropertiesAsync(TwinCollection reportedProperties);
+        Task UpdateReportedPropertiesAsync(ClientTwinProperties reportedProperties);
```
Comment (drwill-ms) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/IModuleClient.cs:26]:
```suggestion
        Task UpdateReportedPropertiesAsync(ReportedProperties reportedProperties);
```

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/ModuleClient.cs [hunk 2]
```diff
@@ -96,7 +96,7 @@ public async Task<Twin> GetTwinAsync()
             }
         }
 
-        public async Task UpdateReportedPropertiesAsync(TwinCollection reportedProperties)
+        public async Task UpdateReportedPropertiesAsync(ClientTwinProperties reportedProperties)
```
Comment (drwill-ms) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/ModuleClient.cs:99]:
```suggestion
        public async Task UpdateReportedPropertiesAsync(ReportedProperties reportedProperties)
```

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/RetryingServiceClient.cs [hunk 3]
```diff
@@ -56,7 +55,7 @@ class DeviceClientRetryStrategy : ITransientErrorDetectionStrategy
             static readonly ISet<Type> NonTransientExceptions = new HashSet<Type>
             {
                 typeof(ArgumentException),
-                typeof(UnauthorizedException)
+                typeof(IotHubServiceException)
             };
 
             public bool IsTransient(Exception ex) => !NonTransientExceptions.Contains(ex.GetType());
```
Comment (drwill-ms) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/RetryingServiceClient.cs:61]: Is if you can refactor to ex is `IotHubServiceException hubEx && hubEx.IsTransient`

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/ServiceClient.cs [hunk 4]
```diff
@@ -9,37 +9,37 @@ namespace Microsoft.Azure.Devices.Edge.Agent.IoTHub
 
     public class ServiceClient : IServiceClient
     {
-        readonly RegistryManager rm;
+        readonly IotHubServiceClient serviceClient;
         readonly string deviceId;
 
         public ServiceClient(string deviceConnectionString, string deviceId)
```
Comment (drwill-ms) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/ServiceClient.cs:15]: rename the pram for them

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/sdkClient/SdkModuleClientProvider.cs [hunk 5]
```diff
@@ -6,13 +6,13 @@ namespace Microsoft.Azure.Devices.Edge.Agent.IoTHub.SdkClient
 
     public class SdkModuleClientProvider : ISdkModuleClientProvider
     {
-        public ISdkModuleClient GetSdkModuleClient(string connectionString, ITransportSettings settings)
+        public ISdkModuleClient GetSdkModuleClient(string connectionString, IotHubClientTransportSettings settings)
         {
             ModuleClient moduleClient = ModuleClient.CreateFromConnectionString(connectionString, new[] { settings });
```
Comment (drwill-ms) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/sdkClient/SdkModuleClientProvider.cs:11]: Refactor to remove the transport settings array

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/sdkClient/WrappingSdkModuleClient.cs [hunk 6]
```diff
@@ -45,9 +43,9 @@ public Task SetMethodHandlerAsync(string methodName, MethodCallback callback)
         public Task SetDefaultMethodHandlerAsync(MethodCallback callback)
             => this.sdkModuleClient.SetMethodDefaultHandlerAsync(callback, null);
 
-        public Task<Twin> GetTwinAsync() => this.sdkModuleClient.GetTwinAsync();
+        public Task<TwinProperties> GetTwinAsync() => this.sdkModuleClient.GetTwinAsync();
 
-        public Task UpdateReportedPropertiesAsync(TwinCollection reportedProperties)
+        public Task UpdateReportedPropertiesAsync(ClientTwinProperties reportedProperties)
```
Comment (drwill-ms) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/sdkClient/WrappingSdkModuleClient.cs:48]: same

File: edge-modules/functions/binding/src/Microsoft.Azure.WebJobs.Extensions.EdgeHub/EdgeHubAsyncCollector.cs [hunk 7]
```diff
@@ -41,16 +43,16 @@ public EdgeHubAsyncCollector(EdgeHubAttribute attribute)
...
         /// <param name="cancellationToken">a cancellation token. </param>
         /// <returns>Task</returns>
-        public async Task AddAsync(Message message, CancellationToken cancellationToken = default(CancellationToken))
+        public async Task AddAsync(TelemetryMessage message, CancellationToken cancellationToken = default(CancellationToken))
         {
-            byte[] payload = message.GetBytes();
-            Message copy = Utils.GetMessageCopy(payload, message);
-            IList<Message> batch = null;
+            byte[] payload = SerializeToByteArray(message.Payload);
+            TelemetryMessage copy = Utils.GetMessageCopy(payload, message);
```
Comment (drwill-ms) [inline edge-modules/functions/binding/src/Microsoft.Azure.WebJobs.Extensions.EdgeHub/EdgeHubAsyncCollector.cs:54]: Look at `TelemetryMessage.Clone()`.

File: edge-modules/functions/binding/src/Microsoft.Azure.WebJobs.Extensions.EdgeHub/EdgeHubTriggerBindingProvider.cs [hunk 8]
```diff
@@ -68,7 +68,7 @@ async Task TrySetEventDefaultHandlerAsync()
             await this.moduleClient.SetMessageHandlerAsync(this.FunctionsMessageHandler, null);
         }
 
-        async Task<MessageResponse> FunctionsMessageHandler(Message message, object userContext)
+        async Task<MessageAcknowledgement> FunctionsMessageHandler(IncomingMessage message, object userContext)
```
Comment (drwill-ms) [inline edge-modules/functions/binding/src/Microsoft.Azure.WebJobs.Extensions.EdgeHub/EdgeHubTriggerBindingProvider.cs:71]: remove userContext param

File: edge-modules/functions/binding/src/Microsoft.Azure.WebJobs.Extensions.EdgeHub/Utils.cs [hunk 9]
```diff
@@ -8,9 +8,9 @@ namespace Microsoft.Azure.WebJobs.Extensions.EdgeHub
 
     static class Utils
     {
-        public static Message GetMessageCopy(byte[] payload, Message message)
+        public static TelemetryMessage GetMessageCopy(byte[] payload, TelemetryMessage message)
```
Comment (drwill-ms) [inline edge-modules/functions/binding/src/Microsoft.Azure.WebJobs.Extensions.EdgeHub/Utils.cs:11]: See why they use a byte[] for payload and see if they are unnecessarily encoding and serializing.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/RetryingServiceClient.cs [hunk 10]
```diff
@@ -56,10 +55,22 @@ class DeviceClientRetryStrategy : ITransientErrorDetectionStrategy
             static readonly ISet<Type> NonTransientExceptions = new HashSet<Type>
             {
                 typeof(ArgumentException),
```
Comment (drwill-ms) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/RetryingServiceClient.cs:57]: Might add `OperationCanceledException` to this list. Also, `InvalidOperationException`.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/sdkClient/WrappingSdkModuleClient.cs [hunk 11]
```diff
@@ -23,7 +21,7 @@ public Task OpenAsync()
             }
             catch (Exception)
             {
-                this.sdkModuleClient?.Dispose();
+                this.sdkModuleClient?.DisposeAsync();
```
Comment (drwill-ms) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/sdkClient/WrappingSdkModuleClient.cs:24]:
```suggestion
                await this.sdkModuleClient.DisposeAsync();
```

**Review by drwill-ms — APPROVED** [review]

File: edge-modules/functions/binding/src/Microsoft.Azure.WebJobs.Extensions.EdgeHub/Utils.cs (hunk 9 above)
Comment (brycewang-microsoft) [reply edge-modules/functions/binding/src/Microsoft.Azure.WebJobs.Extensions.EdgeHub/Utils.cs]: leaving this comment open as I need to double check the usage of method `GetMessageCopy()` with the edge team

**Pull request merged**